components/filter/filter.js

```diff
@@ -0,0 +1,270 @@
+import { getFirstFocusableDescendant } from '../../helpers/focus.js';
+import { LocalizeCoreElement } from '../../lang/localize-core-element.js';
+import { RtlMixin } from '../../mixins/rtl-mixin.js';
+
+/**
+ * A filter component that contains one or more dimensions a user can filter by.
+ * This component is in charge of all rendering.
+ * @slot - Dimension components used by the filter to construct the different dimensions locally
+ * @fires d2l-filter-change - Dispatched when a dimension's value(s) have changed
+ */
+class Filter extends LocalizeCoreElement(RtlMixin(LitElement)) {
+
+	static get properties() {
+		return {
+			/**
+			 * Disables the dropdown opener for the filter
+			 */
+			disabled: { type: String, reflect: true },
+			_activeDimensionKey: { type: String },
+			_dimensions : { type: Array }
+		};
+	}
+
+	static get styles() {
+		return [bodyStandardStyles, css`
+			div[slot="header"] {
+				padding: 18px 6px;
+			}
+			.d2l-filter-dimension-header {
+				align-items: center;
+				display: flex;
+				width: 100%;
+			}
+			.d2l-filter-dimension-header-text {
+				padding-right: calc(2rem + 2px);
+				text-align: center;
+				width: 100%;
+			}
+			:host([dir="rtl"]) .d2l-filter-dimension-header-text {
+				padding-left: calc(2rem + 2px);
+				padding-right: 0;
+			}
+
+			.d2l-filter-dimension-set-value-text {
+				overflow: hidden;
+				text-overflow: ellipsis;
+				white-space: nowrap;
+			}
+		`];
+	}
+
+	constructor() {
+		super();
+		this.disabled = false;
+		this._dimensions = [];
+	}
+
+	firstUpdated(changedProperties) {
+		super.firstUpdated(changedProperties);
+
+		this._dimensionsSlot = this.shadowRoot.querySelector('slot');
+
+		// Prevent these events from bubbling out of the filter
+		this.addEventListener('d2l-hierarchical-view-hide-complete', this._stopPropagation);
+		this.addEventListener('d2l-hierarchical-view-hide-start', this._stopPropagation);
+		this.addEventListener('d2l-hierarchical-view-show-complete', this._stopPropagation);
+		this.addEventListener('d2l-hierarchical-view-show-start', this._stopPropagation);
+		this.addEventListener('d2l-hierarchical-view-resize', this._stopPropagation);
+	}
+
+	render() {
+		const header = this._buildHeader();
+		const dimensions = this._buildDimensions();
+
+		return html`
+			<d2l-dropdown-button-subtle
+				@d2l-dropdown-close="${this._handleDropdownClose}"
+				@d2l-dropdown-open="${this._stopPropagation}"
+				@d2l-dropdown-position="${this._stopPropagation}"
+				text="${this.localize('components.filter.filters')}"
+				?disabled="${this.disabled}">
+				<d2l-dropdown-menu min-width="285" max-width="420" no-padding-header>
+					${header}
+					<d2l-menu label="${this.localize('components.filter.filters')}">
+						<slot
+							@d2l-filter-dimension-data-change="${this._handleDimensionDataChange}"
+							@d2l-filter-dimension-slot-change="${this._handleSlotChange}"
+							@slotchange="${this._handleSlotChange}"
+						></slot>
```

Comment:
I guess this slot doesn't need to go inside the menu? Just wondering if it could live... somewhere else so that it isn't confusing menu with children that it doesn't understand.

Reply:
Hmm, good point - I think it was more important in previous iterations trying to keep event flow more natural, but I think I can move it to the bottom

Reply:
Cool, yeah not a big deal but it mostly made me do a double-take... like "Wait, why is there a slot in the menu, aren't we rendering the menu items... oh I see"

Reply:
Yeah I think it would be best to move this slot out of the menu.  `d2l-menu` does wire up to its `slotchange` event, and while it might tolerate non-menu-item elements, it does some work that is ideally avoided.